File: Week02/Mod1-troublesort.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll  long long int
#define vi  vector<int>
#define vll vector<long long int>


int main(){
    
    std::ios::sync_with_stdio(false);
    
    int T;
    cin >> T; 

    for(int tcase = 1; tcase < T+1; tcase++){ 
        
        int N;
        cin >> N;        

        int num = 0;
        vi v; // This stores the input array
        for(int i = 0; i < N; i++){ 
            cin >> num;            
            v.push_back(num);
        }

        // Prepare to split the input array
        // into subarrays with odd and even indices

        vi odd, even;

        for(int i = 0; i < N; i++){
            if(i % 2 == 0)
                even.push_back(v[i]);
            else
                odd.push_back(v[i]);
        }

        // Sort the spliced arrays separately

        sort(even.begin(),even.end());
        sort(odd.begin(),odd.end());

        // Interleave back the two spliced and sorted arrays

        vi vsorted;

        for(int i = 0; i < N; i++){
            if(i % 2 == 0){
                vsorted.push_back(even[0]);
                even.erase(even.begin());
            }
            else{
                vsorted.push_back(odd[0]);
                odd.erase(odd.begin());
            }
        }


        // Check if the combined array is sorted or not
        int flag = -1;
        for(int i = 0; i < N; i++){
            if(i < N-1){
                // This identifies the first pair that is out of order
                if(vsorted[i] > vsorted[i+1]){ 
                    flag = i;
                    // When such a pair is found we break out of this loop
                    break; 
                }
            }
        }

        // Output based on whether the flag tracking inversions 
        // was triggered or not:
        
        if(flag == -1)
            cout << "Case #" << tcase << ": OK\n";
        else 
            cout << "Case #" << tcase << ": " << flag << "\n";
    }
    
}

File: Week02/Mod1-troublesort.py
'''
Problem: Trouble Sort
URL: https://codingcompetitions.withgoogle.com/codejam/round/00000000000000cb/00000000000079cb
Author : Ayush Modi
Discord Tag : hot9cups#3125
'''

tests = int(input())

for test in range(1, tests + 1):
    n = int(input())
    nums = list(map(int, input().split()))
    
    # 1)Splitting the input array into subarrays with odd and even indices
    # 2)Sorting the spliced arrays
    odds = sorted([num for num in nums[1::2]])
    evens = sorted([num for num in nums[::2]])
    finalList = []

    # Interleaving the spliced and sorted arrays
    for num1, num2 in zip(evens, odds):
        finalList.append(num1)
        finalList.append(num2)
    if len(evens) > len(odds):
        finalList.append(evens[-1])
        
    
    badIdx = -1
    # Check if combined array is sorted, if not - store the first index of sorting error in 'badIdx'
    for i in range(len(nums) - 1):
        if(finalList[i] > finalList[i+1]):
            badIdx = i
            break
    
    if badIdx == -1:
        print(f"Case #{test}: OK")
    else:
        print(f"Case #{test}: {badIdx}")